src/cressi_goa.c

```diff
@@ -474,30 +534,84 @@ cressi_goa_device_foreach (dc_device_t *abstract, dc_dive_callback_t callback, v
 		const unsigned char *dive_data = dc_buffer_get_data (dive);
 		size_t dive_size = dc_buffer_get_size (dive);
 
-		// Verify the header in the logbook and dive data are identical.
-		// After the 2 byte dive number, the logbook header has 5 bytes
-		// extra, which are not present in the dive header.
-		if (dive_size < SZ_HEADER - 5 ||
-			memcmp (dive_data + 0, logbook_data + offset + 0, 2) != 0 ||
-			memcmp (dive_data + 2, logbook_data + offset + 7, SZ_HEADER - 7) != 0) {
-			ERROR (abstract->context, "Unexpected dive header.");
-			status = DC_STATUS_DATAFORMAT;
-			goto error_free_dive;
+		if (conf.id_version == 0) {
+			// Verify the header in the logbook and dive data are identical.
+			// After the 2 byte dive number, the logbook header has 5 bytes
+			// extra, which are not present in the dive header.
+			if (dive_size < SZ_HEADER - 5 ||
+				memcmp (dive_data + 0, logbook_data + offset + 0, 2) != 0 ||
+				memcmp (dive_data + 2, logbook_data + offset + 7, SZ_HEADER - 7) != 0) {
+				ERROR (abstract->context, "Unexpected dive header.");
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
+		} else {
+			const unsigned char *current_logbook_entry = logbook_data + offset;
+			/* The dive header formats start as follows:
+			 *
+			 *  ID  | SAMPLES | START | RATE | Dive-type specific data ...
+			 *  u16 |   u16   |  DATE |  u8  | ...
+			 *
+			 *  START = A Date in the "default" byte-wise form u16u8u8u8u8 for YMDhm
+			 *  RATE = SampleRate - 1=0.5sec 2=1sec 3=2sec 4=5sec
+			 */
+			unsigned short log_entry = array_uint16_le(current_logbook_entry);
+			unsigned short log_entry_ = array_uint16_le(dive_data);
+			if (log_entry_ != log_entry) {
+				ERROR (abstract->context, "Unexpected log entry %d != %d.", log_entry_, log_entry);
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
+			const unsigned char *start_date = current_logbook_entry + 3;
+			if (memcmp(dive_data + 4, start_date, 6) != 0) {
+				unsigned long long is = array_uint64_be(dive_data + 4);
+				unsigned long long should = array_uint64_be(start_date);
+				is >>= 16;
+				should >>= 16;
+				ERROR (abstract->context, "Unexpected start date 0x%0llx != 0x%0llx.", is, should);
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
 		}
 
-		// Those 5 extra bytes contain the dive mode, which is required for
-		// parsing the dive data. Therefore, insert all 5 bytes again. The
-		// remaining 4 bytes appear to be some 32 bit address.
-		if (!dc_buffer_insert (dive, 2, logbook_data + offset + 2, 5)) {
-			ERROR (abstract->context, "Out of memory.");
+		/* Construct a package of form
+		 * [0] Header
+		 * [1] ID
+		 * [2] logbook entry
+		 * [3] dive data
+		 * in order to be able to re-process downloaded data at a later point in time.
+		 */
+
+		/* Inject the header
+		 * [0] = divecomputer  -  0xdc
+		 * [1] = divecomputer  -  0xdc
+		 * [2] = pkg version   -  0x**
+		 * [3] = ID-length     -  0x**
+		 */
```

Comment:
I'm not sure whether a real header is necessary here. Libdivecomputer does not really define the data format. It's the dive computer manufacturer that does. We're only packing a few of those blobs together. Only the length of the ID blob is really necessary because that one doesn't have a fixed length.

Regarding the logbook entry, I'm not really sure whether we should keep that entirely or not. Most of the data in there is also present in the dive header.

For the old format, only the 1 byte dive mode, and 4 byte address is different. The address (of the header in the flash memory) is not very important, so that leaves only the dive mode byte.

For the new format, we have the 1 byte dive mode again, a 2 byte version field, and a 4 byte length field. The rest is also present in the dive header. The length field is the total size of the dive, and we already have that. No idea what the version field is used for. It contains the value 1 for all your dives.

Let's keep it for now, and I'll come back to this, once I also reviewed the parser parts.

Reply:
I'm just thinking about the following use case: I want to use `dctool download -f RAW` to backup all my dives. This backup should be in the format that it came from the DC, in order to be able to parse it again at a later point in time.

Once we've merged this PR there will already be two different RAW formats for the Cressi goa class. The way I've implemented it now is to be able to handle a potential third format, because we can not be sure what the vendors will do in the future. Maybe it's also gonna be YAGNI. For sure we're now able to parse the two formats.

> We're only packing a few of those blobs together. Only the length of the ID blob is really necessary because that one doesn't have a fixed length.

Exactly - and to be able to distinguish how we packed them together I've added the first three bytes. Strictly speaking is the `ID-length` already "`libdivecomputer-cressi-goa-v0`-specific".

IMO if we'd want to be really correct the header should look like:

```
[0] = divecomputer  -  0xdc
[1] = divecomputer  -  0xdc
[2] = pkg version   -  0x**
[3..4] = device     -  0x**** -- a "manufacturer/device" identifier
```

like that `dctool parse` wouldn't even require a `-d <foo>` option in the future.

> Regarding the logbook entry, I'm not really sure whether we should keep that entirely or not. Most of the data in there is also present in the dive header.

I simply did what we agreed on upfront in our mail exchange on 2024-04-25 09:04 `[...] pre-pend the logbook entry to the dive header [...]`. I also think that this is the best way to go. It's cheap and gives us all options later.

> No idea what the version field is used for. It contains the value 1 for all your dives.

My guess is that Cressi also realized that it's easier to include a version field to indicate how the output looks, instead of having to guess.

v4 introduced two version fields: 
1. in the reply to the `VERSION` command, now indicating the "API version" (`4`)
2. in each Logbook entry, now indicating the "dive header format version" (`1`).

I've added comments in the latest commit on how the "logbook" and "dive header format" look like in detail for the different known versions.

Reply:
> That means the libdivecomputer parser can also be used to import those files.

But that's never going to be the case for this class of DC, since the raw dive data is incomplete and we require (at least parts of) both ID and logbook to be able to parse the raw data.

> In hindsight, that was a mistake of course.

Meh, I'm not sure. "It works"^TM and that's what counts IMO. Better have something imperfect that works for the moment and gets improved later when one knows more, than trying to achieve the perfect solution that will never be finished.

> I still need to have another look at how the new minimal header should look like.

For the Cressi goa class we need the following list of parameters to be able to parse the dive data:

```
apiversion - <u16> -- 0 ... 4
divemode - <u8> -- 0,1,2,3,5
fingerprint - <u48>
samplenum - <u16>
samplerate - <u8> -- 0 ... 4
```

The fingerprint is not strictly necessary, but it should be kept.

> ```
> [0] = divecomputer  -  0xdc
> [1] = divecomputer  -  0xdc
> [2] = pkg version   -  0x**
> [3..x] = device id  -  0x**** -- a "manufacturer/device" identifier
> ```

If we chose that or a similar format the user could even forget about the DC this data originates from, since it is contained within.

One could also argue that a simplified container format that only has the necessary information (the parameters as listed above) is sufficient, but I don't think that this brings any real advantage over having everything (complete ID, header and raw dive data) in there.

> I'm not really looking forward to maintain support for this mistake forever, so I think I'll bite the bullet on this one and drop support for the current format

Sounds like a plan.

> For example Uwatec/Scubapro does that, and Shearwater also supports exporting the RAW dive data

I guess we'll have to evaluate this based on each Manufacturer. One is nice and provides a sane export-format, the next makes our life hard. The container above at least allows us to identify "our format"^TM.